### PR TITLE
Update storage

### PR DESCRIPTION
**Pull request by amityadav0**
- Verify we have only single data source
- Set clear boundaries for `ctx` and `store`
- Fix tests

### CODE REVIEW

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:71]:
```suggestion
   * Queries account by accountIndex or `ctx.tx.snd` (if `accountIndex==0`).
   * Throws exception if account is not found.
```

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:20]:
```suggestion
   * All the functions query or update only a state copy from the interpreter, not the `runtime.store`.
```

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:108]:
```suggestion
   * Upserts app global state.
   * Throws error if app is not found.
   * @param appId: application id
   * @param key: app global state key
   * @param value: value associated with a key
```

**Review by robert-zaremba — COMMENTED** [review]
Good catch - interpreter state should be used rather than a runtime one.

File: packages/runtime/src/interpreter/interpreter.ts
Comment (amityadav0) [reply packages/runtime/src/interpreter/interpreter.ts:19]: ok, changing

File: packages/runtime/src/interpreter/interpreter.ts
Comment (amityadav0) [reply packages/runtime/src/interpreter/interpreter.ts:41]: ok

File: packages/runtime/src/interpreter/interpreter.ts
Comment (amityadav0) [reply packages/runtime/src/interpreter/interpreter.ts:58]: we need `undefined` here because for some opcodes if app is not defined they push 0 to stack

Comment (ratik21) [reply packages/runtime/src/interpreter/interpreter.ts:58]: [`SSCAttributesM`](https://github.com/scale-it/algorand-builder/blob/master/packages/runtime/src/types.ts#L86) represents [`SSCAttributes`](https://github.com/scale-it/algorand-builder/blob/master/packages/types-algosdk/index.d.ts#L549) from algosdk, with one stipulation: we are using `maps` instead of `arrays` ( in `'global-state'` here). 
Added the suffix `M` as we are using maps. We can rename to something more meaningful if not clear.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:88]:
```suggestion
  assertAppDefined (app?: SSCAttributesM, appId: number, line?: number): SSCAttributesM {
```
We are using `| undefined` in 1st arg and `?` in 3rd

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:107]: Please create a function `getApp` in `StoreAccount`. We should operate on an account attribute (`createdApps` in this case) in account only. So `runtime.getApp() => calls account.getApp()` which get's the app from account's global state.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:129]: Same as above. `runtime.getGlobalState => calls account.getGlobalState`. We can check the getLocalState function below (I think we discussed these updates in the `improve DevUx` task, but at that time we were operating on `globalApps` only and not `store.accounts`.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:24]: :+1:

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:137]: Should we remove this function from runtime? Is it possible that during testing user wants to set global state explicitly before calling the smart contract?

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:157]: We have `initializeAccounts()` in runtime constructor. But currently we create an app/asset after runtime initialization. eg.
```
const runtime = new Runtime([john, bob]); // sets this.store.accounts
runtime.addApp({..}), runtime.updateApp({..})..etc afterwards. 
```
Since we don't have any created apps or assets during initialization of runtime, do lines 151-153, 155-157 ever get executed?

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:234]: Please don't operate on `ctx` in Runtime. We are directly contradicting the note above the runtime class here. Let's be consistent with our definition: `interpreter operates on ctx (using deepCopy of store) and runtime operates on store`. 

So, we can do these updates on `runtime.store` (if needed), and then assign a deepCopy to `ctx` before `run`. eg.
```
this.store.accounts.set(senderAcc.address, senderAcc);
this.store.globalApps.set(app.id, senderAcc.address);
this.ctx.state = deepCopy(this.store) // assign a deep copy here instead of setting directly
await this.run(..)
..
```

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:248]: Same as above. Also, do we need this line (keeping `ctx` after transaction is executed) ? I mean `ctx` will be overwritten on `runtime.executeTx` and in `Runtime` we are using `store` to getAccount.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:363]: is `this.assertAccountDefined(this.store.accounts.get(accountAddr))` returning a deepCopy ?

File: packages/runtime/test/integration/deleteApp.ts
Comment (ratik21) [inline packages/runtime/test/integration/deleteApp.ts:69]:
```suggestion
    // verify app is not deleted - using getApp function
```

File: packages/runtime/test/src/interpreter/opcode-list.ts
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:2348]: Should we use `runtime.getLocalState` (Assuming user will do the same)?

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:107]: yes

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:137]: I don't think that's ideal, User will be testing their TEAL files, not our code. If that makes sense

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:157]: yes, they do, when we call the constructor new Runtime([john, bob]). and if john and bob, have createdAssets or createdApps, we are mocking these.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:234]: we cannot do that, for deployer functions(addApp, updateApp).
because what if program gets failed, you would have updated the store already.
but according to spec, we cannot do that, first ctx will be updated, then only we move to store

also this is similar to what we do in `run` first we set the ctx and then we pass it to interpreter, and if everything goes correct, we update store.(it is more clear and avoids bug)
I can add a detailed doc for more clarification.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:363]: yes, similar to addApp, did you figure what was going on with addAPp ?

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:248]: same as above, yes, we don't need it.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:234]: One other way would be moving appApp, updateApp to interpreter.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:88]: yes

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:234]: hmm, then we need to figure something out. The current implemention seems incorrect by definition.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:363]: When i tested `addApp` it wasn't returning a deepCopy. That's why i asked here

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:157]: how can john or bob create new asset or app before initializing runtime?

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:234]: > because what if program gets failed, you would have updated the store already.

Agreed. The state would be updated even if program fails. But i see in the below code, we are deleting this app from the account (since this is a zero indexed app, we only need it's attribute and set new state)

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:157]: If you see we have mocks, which has hard coded createdAssets, this is used for that,
I haven't structured it like that, I think it was added by you. We can update in new PR

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:234]: One way would be to pass ctx to intrpreter.addApp and let it handle this and return updated ctx
Which we can update in runtime.addApp

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:157]: I think it's good for user, user can send updated states if he wishes.
We can discuss it's pros and cons.

Comment (ratik21) [reply packages/runtime/src/runtime.ts:157]: I wasn't saying to remove it, was rather curious when this does gets executed.
For mocks the structure seems fine, as we set in globalassets and App directly there. From users perspective, I only see this code getting executed when the user explicitly modifies createdAssets or Map by using account.createdApps.set(..)..

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:363]: verified it, working after removing it, will have a look at each function for this too.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:234]: > > because what if program gets failed, you would have updated the store already.
> 
> Agreed. The state would be updated even if program fails. But i see in the below code, we are deleting this app from the account anyway (since this is a zero indexed app, we only need it's attribute and set new state with id appCounter + 1). Will check more

if program fails, error will be thrown therefore we won't reach that line. store will be updated.

Comment (amityadav0) [reply packages/runtime/src/runtime.ts:231]: also i don't think we can move it to interpreter,, that will create cycle. best would be to add doc. what do you think?

File: packages/runtime/src/account.ts
Comment (robert-zaremba) [inline packages/runtime/src/account.ts:90]:
```suggestion
   * Queries app global state value. Returns `undefined` if the key is not present.
```

File: packages/runtime/src/account.ts
Comment (robert-zaremba) [inline packages/runtime/src/account.ts:103]:
```suggestion
   * Queries application by application index. Returns undefined if app is not found.
```

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:21]:
```suggestion
   */
   
```

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:41]:
```suggestion
   * Queries ASA Definitions data by assetID.  Returns undefined if ASA is not deployed.
```

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:55]: Update the second part: explain when the app returns undefined and when it throws exception.

File: packages/runtime/src/interpreter/interpreter.ts
Comment (robert-zaremba) [inline packages/runtime/src/interpreter/interpreter.ts:91]:
```suggestion
   * Queries application by application index. Returns undefined if app is not found.
```

**Review by robert-zaremba — APPROVED** [review]
Pre-approving. Left comments and suggestions. Please also wait for @ratik21  approval to merge.

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:137]: It just seems to me we're taking control away from the user. Eg For a stateful counter application, the user can use `runtime.setGlobalState('counter', 5)` explicitly, and then call the smart contract to assert if counter has incremented by 1 or not. What's the benefit of removing it? 
Anyway, if you will remove it, please update the `/docs` as well. It is mentioned there user can use `setGlobalState`.

File: packages/runtime/src/account.ts
Comment (ratik21) [inline packages/runtime/src/account.ts:94]: Just to confirm, we haven't removed this function from `runtime` right? `runtime.getGlobalState` will call `account.getGlobalState`. Right?

Comment (amityadav0) [reply packages/runtime/src/account.ts:94]: yes

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:231]: > if program fails, error will be thrown therefore we won't reach that line. store will be updated.

We can wrap the `run` in try {} catch () and throw error after deleting ?
Anyways, adding doc seems to be OK here. Please explain why are operating on `ctx` here in a doc string.

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:137]: fair enough, restoring setGlobalState

File: packages/runtime/src/runtime.ts
Comment (ratik21) [inline packages/runtime/src/runtime.ts:244]: Also, since `this.assertAccountDefined(this.store.accounts.get(sender.addr))` is not returning a deepCopy, do we need this line?

Comment (amityadav0) [reply packages/runtime/src/runtime.ts:244]: ok, removing it, please make sure tests are good for this one.

**Review by ratik21 — APPROVED** [review]
Please check `/docs/testing-teal.md` as well before merging

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:244]: checking

File: packages/runtime/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/src/runtime.ts:231]: ok

File: packages/runtime/src/runtime.ts
Comment (ratik21) [reply packages/runtime/src/runtime.ts:244]: > ok, removing it, please make sure tests are good for this one.

Please check at other places as well

Comment (ratik21) [reply packages/runtime/src/runtime.ts]: Seems ok

**Pull request merged**
